### PR TITLE
Clean up LSClockState

### PR DESCRIPTION
**Pull request by ryanrhee**
- pull out methods that don't need `self` into static functions
- use NSFastEnumeration instead of C for-loops
- wrap the update call in init to run in `_sampleQueue`

### CODE REVIEW

**Comment by ryanrhee** [comment]
CI seems to have failed because the submodule update failed:

```
$ git submodule update --init --recursive
Submodule 'lightstep-tracer-common' (git@github.com:lightstep/lightstep-tracer-common.git) registered for path 'lightstep-tracer-common'
Cloning into '/Users/travis/build/lightstep/lightstep-tracer-objc/lightstep-tracer-common'...
Warning: Permanently added 'github.com,192.30.253.112' (RSA) to the list of known hosts.
Permission denied (publickey).
fatal: Could not read from remote repository.
```

I imagine CI isn't configured correctly.

File: Pod/Classes/LSClockState.m
Comment (bcronin) [inline Pod/Classes/LSClockState.m:101]: Hi @ryanrhee, apologies in advance if my comments below are inaccurate. It's been a while since I've been working with Objective-C code :)

I believe the problem _might_ be in the async callback created on line 91 before (101 after). The `dispatch_async` callback is referencing `self` but there are no guarantees around when the dispatch may occur and, by extension, no guarantees that `self` will still be a valid reference inside the callback.  The pattern that I suspect should be used here would be something along the lines of what's discussed on this [stackoverflow post](https://stackoverflow.com/questions/21987067/using-weak-self-in-dispatch-async-function) or here in [this repo](https://github.com/lightstep/lightstep-tracer-objc/blob/34cffe7341aaf5614f7232762b732e482fd32885/Pod/Classes/LSTracer.m#L296). Looks like this callback got added [about a year ago](https://github.com/lightstep/lightstep-tracer-objc/blame/master/Pod/Classes/LSClockState.m#L91).

If that theory is correct, then `self.samples.copy` is an indirection of a garbage pointer, which makes the pointer passed to `_persistSamplesToUserDefaults` invalid, and eventually may explain the call stacks. I believe the fix could be to add a `__weak` pointer outside the `dispatch_async` and check the strong reference is non-null before proceeding in the callback.

Comment (ryanrhee) [reply Pod/Classes/LSClockState.m:101]: Hi @bcronin !

> dispatch_async callback is referencing self

Yup. And, because this block isn't using using a weak reference (i.e. `__weak typeof(self) weakSelf = self;`) the block should retain `self`.

> but there are no guarantees around when the dispatch may occur

Correct

> and, by extension, no guarantees that self will still be a valid reference inside the callback

I don't think this is true in an ARC environment, since, as I mentioned in the comment earlier, the block will keep a reference to `self`. While this is often discouraged as it leads to retain cycles, in this case, this helps, since it ensures that `self` won't be deallocated until the block executes.

Of course, if ARC is explicitly disabled in this project, you'd be correct. I don't think that's the case, though - if so, we'd see things like `[foo retain]` and `[foo release]` everywhere.

(You misspelled my name / github handle, FYI - & no worries.)

Comment (bcronin) [reply Pod/Classes/LSClockState.m:101]: Thanks for the clarification. Ah, I knew there was something unusual around the strong reference, but had it backwards (retain cycle, not a potential deallocation). It _has_ been a while since I've written any Objective-C :)

**Comment by bcronin** [comment]
The changes in the code looks correct and fine, however admittedly it's a little unclear to me what the functional difference is. 

The loop style changes seem to be purely code style clean-up (which is great, but just want to confirm there's no functional change).

Moving `_persistSamplesToUserDefaults` to a static function also seems to be functionally equivalent to the prior code since the immediate caller references `self`

The final change is to move the `init` call to `_update` into a `dispatch_sync` call. I don't fully understand the desired outcome of this change. Based on my understanding, it means the `_update` will happen serially since it's in a queue now and it *may* (?) happen in a different thread but not necessarily (["As an optimization, this function invokes the block on the current thread when possible."](https://developer.apple.com/documentation/dispatch/1452870-dispatch_sync)). Since this is in the `init` function, it seems like there's nothing that could "skip in line" so it's not clear to me what the effect of putting it in the `_samplesQueue` is versus running it directly.

Aside: the changes in your PR aside, do you see any concerns with thread-safety in this code? I'm a rusty on the details of GCD, but if the `_sampleQueue` functions are potentially run in a different thread than the any access to `samples` (which is accessible outside the class) then it seems like there _could_ be concurrent access to an `NSMutableArray`. If that's the case, then it seems possible that the input to `persistSamplesToUserDefaults` could be corrupted.

Apologies for the many questions here, but as mentioned, my Obj-C is rusty and I'm struggling a bit to pinpoint the desired changes in the code execution beyond the code style simplification.

**Comment by ryanrhee** [comment]
> The changes in the code looks correct and fine, however admittedly it's a little unclear to me what the functional difference is.

With the exception of wrapping one of the functions in the initializer inside a `dispatch_sync`, all other changes are mostly cosmetic in nature. (The for-loop style change will likely have a small perf improvement, but that's not the reason for the change.)

> The final change is to move the init call to _update into a dispatch_sync call. I don't fully understand the desired outcome of this change. Based on my understanding, it means the _update will happen serially since it's in a queue now and it may (?) happen in a different thread but not necessarily ("As an optimization, this function invokes the block on the current thread when possible."). Since this is in the init function, it seems like there's nothing that could "skip in line" so it's not clear to me what the effect of putting it in the _samplesQueue is versus running it directly.

It's to make sure all access to `_samples` happens on the `_samplesQueue` thread.

> Aside: the changes in your PR aside, do you see any concerns with thread-safety in this code? I'm a rusty on the details of GCD, but if the _sampleQueue functions are potentially run in a different thread than the any access to samples (which is accessible outside the class) then it seems like there could be concurrent access to an NSMutableArray. If that's the case, then it seems possible that the input to persistSamplesToUserDefaults could be corrupted.

AFAIK, all outwards functions that access `_samples` or `self.samples` punts out to `_samplesQueue`, so there aren't any obvious thread safety concerns. What I mean is, since the queue is marked as `DISPATCH_QUEUE_SERIAL`, regardless of which thread its blocks are run on, only one will execute at a time.

Let me know if you have any more questions.

**Review by bcronin — APPROVED** [review]
Thanks for the clarifications. This looks good.

**Comment by bcronin** [comment]
(FYI, I'm trying to get the TravisCI issue resolved)

**Comment by ryanrhee** [comment]
CI is passing.

**Review by bcronin — APPROVED** [review]

**Comment by ryanrhee** [comment]
@bcronin FYI i can't merge this since I don't have write access.

**Pull request merged**
